### DISCUSSION TITLE
[RESOURCES] Blue Archive Characters Spine Asset

### DISCUSSION THREAD

**Discussion by kiraio-moe** [Tualin14/baPlayer#14] [General]
Hello, Sensei-tachi!

I want to share Blue Archive Spine characters that more complete and well renamed.
You can check it out here: https://github.com/kiraio-moe/Schale-Archive

Well, there's some missing characters (https://github.com/kiraio-moe/Schale-Archive#known-missing-character-spine) though.
If you have them, please make a PR :)

**Comment by ttttttttttttttllllllllllllll** [comment]
I don't currently have the new Spine characters material, but you can get it at https://github.com/xiongnemo/blue-archive-jp-assets-downloader

**Reply by kiraio-moe** [reply]
Ah, it's the latest version anyway.
I get the asset bundle indexes (`.json`) from this https://github.com/xiongnemo/blue-archive-jp-assets-downloader then compare it with the asset that you provided and the latest game client from my friend.

IDK why there's some spine asset that didn't exist in installed game. Even I to try download it using the tool above, no luck.

**Reply by ttttttttttttttllllllllllllll** [reply]
> IDK why there's some spine asset that didn't exist in installed game. Even I try download it using the tool above, no luck.

Because these assets are not in the download link, they should be in the apk.
I'm guessing these other jp version assets should be in the obb file. (I'm currently not looking at it because I have something to do before the 20th, so I'm just guessing.)

The gb version assets are in the preload folder and the game resources folder respectively.

**Reply by kiraio-moe** [reply]
> Because these assets are not in the download link, they should be in the apk.

I doubt it. They already have good Addressables system, why whould they do that to include some particular asset inside the APK?

**Reply by kiraio-moe** [reply]
This list (https://github.com/kiraio-moe/Schale-Archive#known-missing-character-spine) is now has gone down. There are some naming "errors", such as `Zunko` (KR, I guess?) become `Junko` (JP). So I'm confused. As we know, most wikis (GB) prefer to use Japanese names.

The list is now just this NPCs:
- GSC President
- Nagusa
- A.R.O.N.A

As for Nagusa it looks like I've seen it before . Or it's just _DeJavu_? 🤔

**Reply by ttttttttttttttllllllllllllll** [reply]
> I doubt it. They already have good Addressables system, why whould they do that to include some particular asset inside the APK?

Because the assets of the game start tutorial do not need to be downloaded, you can download other assets at the same time as the tutorial. You can find Wakamo, Hasumi and other assets in the start tutorial.

**Reply by kiraio-moe** [reply]
> Because the assets of the game start tutorial do not need to be downloaded, you can download other assets at the same time as the tutorial. You can find Wakamo, Hasumi and other assets in the start tutorial.

I don't think so. Before we first start the game, we are required to download the game data, right?

**Reply by ttttttttttttttllllllllllllll** [reply]
> There are some naming "errors", such as Zunko (KR, I guess?) become Junko (JP). So I'm confused. As we know, most wikis (GB) prefer to use Japanese names.

I think we can use [SchaleDB](https://github.com/lonqie/SchaleDB/blob/main/data/en/students.json) method, with the original name of the game asset for the filename(devName), and make an index(json) to account for the name in different languages(en,jp,cn,kr)

**Reply by ttttttttttttttllllllllllllll** [reply]
> I don't think so. Before we first start the game, we are required to download the game data, right?

I don't konw the jp version because I play global ver. 
In gb ver. It's true. There are in Preload folder and should be downloaded.
But in jp ver, I don't know, so I'm just guessing.

**Reply by kiraio-moe** [reply]
> I think we can use [SchaleDB](https://github.com/lonqie/SchaleDB/blob/main/data/en/students.json) method, with the original name of the game asset for the filename(devName), and make an index(json) to account for the name in different languages(en,jp,cn,kr)

That's a good idea. Wondering how the json structure is?

**Reply by kiraio-moe** [reply]
> I don't konw the jp version because I play global ver. In gb ver. It's true. There are in Preload folder and should be downloaded. But in jp ver, I don't know, so I'm just guessing.

I think the system is the same. Only different servers (and language of course).

**Reply by ttttttttttttttllllllllllllll** [reply]
> That's a good idea. Wondering how the json structure is?

### This is a simple reference

- name

  - jp
  - en
  - zh_cn(or cn)
  - zh_tw(or tw)
  - kr

- path

### like:

```json
[
  {
    "name": {
      "jp": "アル",
      "en": "Aru",
      "cn": "阿露",
      "tw": "亞瑠",
      "kr": "아루"
    },
    "path": "Schale-Archive/Spine 2D/Aru/"
  },
 ...
]
```

You can add more keys if you need more.

**Reply by kiraio-moe** [reply]
> ```json
> [
>   {
>     "name":
>     {
>       "jp": "アル",
>       "en": "Aru",
>       "cn": "阿露",
>       "tw": "亞瑠",
>       "kr": "아루"
>     },
>     "path": "Schale-Archive/Spine 2D/Aru/"
>   },
> ]
> ```

I think, I will add each asset bundle hash to perform some checking. Anyway, thanks~

**Comment by ttttttttttttttllllllllllllll** [comment]
> I think the system is the same. Only different servers (and language of course).

It's api different. 

[jp](https://prod-clientpatch.bluearchiveyostar.com/r52_1_27_dzgkgjtji98s6evee801/iOS/bundleDownloadInfo.json) - u can find `assets-_mx-spinecharacters-wakamo_mask_spr`, but can't find `assets-_mx-spinecharacters-wakamo_spr`

[gb](https://ba.dn.nexoncdn.co.kr/com.nexon.bluearchive/062a8b0da0414731/resource-data.json) - The gb ver. download `Preload` first at game start. The `assets-_mx-spinecharacters-wakamo_spr` in `PreLoad`.

**Comment by ttttttttttttttllllllllllllll** [comment]
I noticed in your project that it says

> First, you need Spine Editor at least v3.8.x and later because Blue Archive character spine is produced in Spine Editor v.4.3.8.96.

I want to know which character spine is produced in Spine Editor v.4.
At the time I started the as project, they were produced in ver. 3.8

**Reply by kiraio-moe** [reply]
> I want to know which character spine is produced in Spine Editor v.4. At the time I started the as project, they were produced in ver. 3.8

Yes, Spine assets produced vary from 3.8 ~ 4.0. Wait a minute, I also forgot which character it was.

**Reply by kiraio-moe** [reply]
Oops, I made a mistake 😅 

I do quick `grep`

![image](https://user-images.githubusercontent.com/58289710/226166376-23d8f550-d76f-473b-9817-f1706fc210d1.png)

and try to open one in Hex Editor.

![asset-spine-ver](https://user-images.githubusercontent.com/58289710/226166673-c35dd21f-52ab-49f2-a689-86967f1e5eef.png)

When you say `3.8`, it immediately clicked in my mind that `4` (red) is not part of `3.8.96`.
Also `3.8.96` is wrong too, the correct version is `3.8.9`.

I will change that in the future.

**Reply by ttttttttttttttllllllllllllll** [reply]
Spine versions are not compatible, so if spine.4 of the game assets exists, we need to add the spine-runtime.4, and the latest version of spine is 4.1.

So we can solve the problem with spine-runtime.3.8 (except 3.8.75, This is also the reason for the related file uploads that I have ignored for now. #10 )

**Reply by kiraio-moe** [reply]
> Spine versions are not compatible, so if spine.4 of the game assets exists, we need to add the spine-runtime.4, and the latest version of spine is 4.1.

I thinks that's pretty useless if the devs uses 2 Spine version, knowing that Spine has compatibility issue.

> So we can solve the problem with spine-runtime.3.8 (except 3.8.75, This is also the reason for the related file uploads that I have ignored for now. #10 )

The funny things when I working in this project for the first time is that, I'm facing the same issue like #10 then I try to delete existing Spine folder and re-import them from <https://esotericsoftware.com/files/runtimes/unity/spine-unity-3.8-2021-11-10.unitypackage>. After that, it gives me this errors even though I'm using the same spine-unity version.

![image](https://user-images.githubusercontent.com/58289710/226173736-005d05d5-1947-4be4-89e8-29c0e57a5fc1.png)

The errors is valid. There's no `CreateSkeletonDataAsset` method in `SkeletonDataAsset` class.

![image](https://user-images.githubusercontent.com/58289710/226173928-5f24e8ad-0208-4709-be02-e67e08a4a6c3.png)

I'm still not familiar with spine-unity yet, so IDK what those line actually doing and unable to fix that.

I reverted back the changes then re-import again, **BUT** only importing those missing class in #10. I look `SkeletonDataAsset` class again and found `CreateSkeletonDataAsset` method.

![image](https://user-images.githubusercontent.com/58289710/226174396-6733d7be-0ffe-4bbf-89d2-7df8b28fecb0.png)

Did you customize `SkeletonDataAsset` class?

**Reply by ttttttttttttttllllllllllllll** [reply]
> Did you customize SkeletonDataAsset class?

Maybe, I forgot it, I just remember that I removed the version limitation of 3.8.75.

But I am going to replace it and plan to make some changes to as to deal with [Spine Runtimes license](http://esotericsoftware.com/support)

> To avoid having your software project become encumbered by the Spine Runtimes license, one solution is to make the Spine Runtimes portion of your software an optional module which is licensed under the Spine Runtimes license. This gives your users the choice to opt in to the Spine Runtimes license if desired, while still allowing you to license the rest of your project as you like.

**Reply by ttttttttttttttllllllllllllll** [reply]
Yes, I wrote this method in there because it can only load [json](https://zh.esotericsoftware.com/spine-unity#%E5%B0%86Skeleton%E6%B7%BB%E5%8A%A0%E5%88%B0%E5%9C%BA%E6%99%AF%E4%B8%AD).
Although it says it can load bytes, it doesn't look right from the official related code, is it official to let read the binary file in text way?
I'm testing whether this approach will work.

So I added this method a few months ago that can read bytes and I modified SkeletonDataAsset.cs

**Reply by ttttttttttttttllllllllllllll** [reply]
> Although it says it can load bytes, it doesn't look right from the official related code, is it official to let read the binary file in text way?
> I'm testing whether this approach will work.

Ok，it's didn't work. So now I still have to customize the spine-runtimes.

**Reply by kiraio-moe** [reply]
Sorry for the late response.

>> Although it says it can load bytes, it doesn't look right from the official related code, is it official to let read the binary file in text way?
>> I'm testing whether this approach will work.
>
> Ok，it's didn't work. So now I still have to customize the spine-runtimes.

Yes, it will not work because it is impossible to load an external binary file as a TextAsset in Unity at runtime.
The TextAsset class is designed to load text-based assets, such as .txt, .json, or .xml files, which are stored in the Unity project's Assets folder.

> Ok，it's didn't work. So now I still have to customize the spine-runtimes.

We can still set the `skeletonData` and `stateData` without putting these method...

```cs
public static SkeletonDataAsset CreateSkeletonDataAsset(SkeletonData skeletonData, AnimationStateData stateData)
{
    SkeletonDataAsset skeletonDataAsset = ScriptableObject.CreateInstance<SkeletonDataAsset>();
    skeletonDataAsset.skeletonData = skeletonData;
    skeletonDataAsset.stateData = stateData;

    return skeletonDataAsset;
}
```

...in `SkeletonDataAsset.cs` by using Reflection. Something like this:

```cs
using System;
using System.Reflection;
using Spine;
using Spine.Unity;
using UnityEngine;

namespace ArisStudio.Utils
{
    public static class SpineHelper
    {
        public static SkeletonDataAsset CreateRuntimeInstance(SkeletonData skeletonData, AnimationStateData stateData)
        {
            // Create a new instance of SkeletonDataAsset
            SkeletonDataAsset skeletonDataAsset = ScriptableObject.CreateInstance<SkeletonDataAsset>();

            // Get the type of SkeletonDataAsset
            Type skeletonDataAssetType = skeletonDataAsset.GetType();

            // Get the skeletonData and stateData fields
            FieldInfo skeletonDataField = skeletonDataAssetType.GetField("skeletonData", BindingFlags.NonPublic | BindingFlags.Instance);
            FieldInfo stateDataField = skeletonDataAssetType.GetField("stateData", BindingFlags.NonPublic | BindingFlags.Instance);

            // Set the values of skeletonData and stateData
            skeletonDataField.SetValue(skeletonDataAsset, skeletonData);
            stateDataField.SetValue(skeletonDataAsset, stateData);

            // Return the SkeletonDataAsset
            return skeletonDataAsset;
        }
    }
}
```

It works perfectly, but I'm not sure if this approach is fine.

**Reply by ttttttttttttttllllllllllllll** [reply]
add this
```cs
            // Set the value of skeletonJSON to let SkeletonDataAsset.cs 114:if (skeletonJSON == null) is false
            skeletonDataAsset.skeletonJSON = new TextAsset("ArisStudio");
```

## It works wonderfully without customize the spine-runtimes. Thank you. ヾ(≧▽≦*)o

```cs
using System;
using System.Reflection;
using Spine;
using Spine.Unity;
using UnityEngine;

namespace ArisStudio.Utils
{
    public static class SpineHelper
    {
        public static SkeletonDataAsset CreateRuntimeInstance(SkeletonData skeletonData, AnimationStateData stateData)
        {
            // Create a new instance of SkeletonDataAsset
            SkeletonDataAsset skeletonDataAsset = ScriptableObject.CreateInstance<SkeletonDataAsset>();

            // Get the type of SkeletonDataAsset
            Type skeletonDataAssetType = skeletonDataAsset.GetType();

            // Get the skeletonData and stateData fields
            FieldInfo skeletonDataField = skeletonDataAssetType.GetField("skeletonData", BindingFlags.NonPublic | BindingFlags.Instance);
            FieldInfo stateDataField = skeletonDataAssetType.GetField("stateData", BindingFlags.NonPublic | BindingFlags.Instance);

            // Set the values of skeletonData and stateData
            skeletonDataField.SetValue(skeletonDataAsset, skeletonData);
            stateDataField.SetValue(skeletonDataAsset, stateData);

            // Set the value of skeletonJSON to let SkeletonDataAsset.cs 114:if (skeletonJSON == null) is false
            skeletonDataAsset.skeletonJSON = new TextAsset("ArisStudio");

            // Return the SkeletonDataAsset
            return skeletonDataAsset;
        }
    }
}
```

**Reply by kiraio-moe** [reply]
> ```cs
>             // Set the value of skeletonJSON to let SkeletonDataAsset.cs 114:if (skeletonJSON == null) is false
>             skeletonDataAsset.skeletonJSON = new TextAsset("ArisStudio");
> ```

Hmm, I didn't have any issues by not putting those line. Can you explain to me why we need it?

> It works wonderfully without customize the spine-runtimes. Thank you. ヾ(≧▽≦*)o

Your welcome :) I suppose we will be use them.

**Reply by ttttttttttttttllllllllllllll** [reply]
> Hmm, I didn't have any issues by not putting those line. Can you explain to me why we need it?

In the original SkeletonDataAsset.cs on line 144. It needs to return false but the skeletonJSON is null with SpineHelper.

```cs
public SkeletonData GetSkeletonData (bool quiet) {
	if (skeletonJSON == null) {
		if (!quiet)
			Debug.LogError("Skeleton JSON file not set for SkeletonData asset: " + name, this);
		Clear();
		return null;
	}
...
}
```

 In the past as version, I commented out [this](https://github.com/Tualin14/ArisStudio/blob/ArisStudio2/Assets/Spine/Runtime/spine-unity/Asset%20Types/SkeletonDataAsset.cs#L123)

**Reply by kiraio-moe** [reply]
> In the original SkeletonDataAsset.cs on line 144. It needs to return false but the skeletonJSON is null with SpineHelper.
> 
> ```cs
> public SkeletonData GetSkeletonData (bool quiet) {
> 	if (skeletonJSON == null) {
> 		if (!quiet)
> 			Debug.LogError("Skeleton JSON file not set for SkeletonData asset: " + name, this);
> 		Clear();
> 		return null;
> 	}
> ...
> }
> ```

Aahh.., it's to make sure there's no error if we call those method. OK ok.

**Comment by kiraio-moe** [comment]
Edit: Where's your comment (‘◉⌓◉’)

That's good ( ╹▽╹ )
Does it work well with CJK characters?

Pada tanggal Sel, 28 Mar 2023 08.05, dzaaaaaa ***@***.***>
menulis:

**Reply by ttttttttttttttllllllllllllll** [reply]
Because I used the [online tool](https://regexr.com/) to verify it, and it caused my judgment error.

It work well. Thank you.
Sorry to have disturbed you. orz

**Comment by ttttttttttttttllllllllllllll** [comment]
You are not in the contributors because I have to delete the git history. So I invited you to become a contributor directly.

**Reply by kiraio-moe** [reply]
No prob.

**Reply by ttttttttttttttllllllllllllll** [reply]
You can change the following readme at will and push it to have your name displayed on the homepage.

**Reply by kiraio-moe** [reply]
> You can change the following readme at will and push it to have your name displayed on the homepage.

There's no need to do it. If I make a commit then merged with main branch, the contributor name will be there.

Ah, yeah. Regarding cleaning up the commit history, in the future development I think it will be wise to `squash commits` that has little changes so we can atleast reduce the history and in the same time preserve the history of what has changed. It will be useful if there's a bug that caused by certain commit so we can easily fix them.

One more thing. If possible, it will be great to use English language in commit title and message so I can more understand the actual context. Because so far, I have been using Google Translate to translate Chinese and of course the results are not very good. I also don't really understand English, so sometimes I also use Google Translate to translate my language into English. 😅

**Reply by ttttttttttttttllllllllllllll** [reply]
> Ah, yeah. Regarding cleaning up the commit history, in the future development I think it will be wise to squash commits that has little changes so we can atleast reduce the history and in the same time preserve the history of what has changed. It will be useful if there's a bug that caused by certain commit so we can easily fix them.

Deleting git history is because some resources were mixed in early, and you can find the repo for downloading ba resources was deleted today.

> If possible, it will be great to use English language in commit title and message so I can more understand the actual context.

The Chinese commit is used because this git commit is not important and I will use English for the critical commit. After all, I am not a native English speaker and am not fluent in English. I don't know how to express ideas in exact words.

**Reply by ttttttttttttttllllllllllllll** [reply]
I use contribute-list and now we can see Contributors in homepage
